libevmasm/ThirdOptimizerTesting.cpp

```diff
@@ -0,0 +1,368 @@
+}
+
+template<typename It, typename = typename std::enable_if<is_assembly_item<decltype(*std::declval<It>())>::value>::type>
+vector<It> parseExpressions(It _begin, It _end, signed _limit = -1, bool _allowNoReturn = false)
+{
+	auto currentEnd = _end;
+	std::vector<It> expressions{};
+	while (currentEnd != _begin && (_limit == -1 || signed(expressions.size()) < _limit))
+	{
+		auto oldEnd = currentEnd;
+		currentEnd = backExpressionBegin(_begin, oldEnd, _allowNoReturn);
+		expressions.push_back(currentEnd);
+		assertThrow(currentEnd >= _begin, OptimizerException, "currentEnd is before begin");
+	}
+
+	return expressions;
+}
+
+template<typename It, typename = typename std::enable_if<is_assembly_item<decltype(*std::declval<It>())>::value>::type>
+vector<It> parseArguments(It _begin, It _end)
+{
+	auto size = _end - _begin;
+	assertThrow(size > 0, OptimizerException, "cannot get arguments from empty _items vector");
+
+	unsigned neededArguments = (_end - 1)->arguments();
+
+	auto expressions = parseExpressions(_begin, _end - 1, neededArguments, false);
+	assertThrow(expressions.size() >= neededArguments, OptimizerException, "more arguments needed than provided");
+
+	expressions = vector<It>(expressions.begin(), expressions.begin() + neededArguments);
+
+	return expressions;
+}
+
+vector<AssemblyItem> createItems(NewOptimizerPattern const& _pattern)
+{
+	if (_pattern.isBound())
+	{
+		auto bound = _pattern.boundItems();
+		return vector<AssemblyItem>(bound.begin, bound.end);
+	}
+
+	switch (_pattern.kind())
+	{
+		case NewOptimizerPattern::Kind::Any:
+			assertThrow(false, OptimizerException, "cannot create items from non-bound pattern");
+		case NewOptimizerPattern::Kind::Constant:
+		{
+			if (_pattern.hasConstant())
+				return {_pattern.constant()};
+			else
+				assertThrow(false, OptimizerException, "cannot create items from non-bound constant pattern without constant value");
+		}
+		case NewOptimizerPattern::Kind::Operation:
+			if (_pattern.hasOperationValues())
+			{
+				auto operands = _pattern.operands();
+
+				vector<AssemblyItem> instructions{};
+
+				for (auto it = operands.rbegin(); it != operands.rend(); it++)
+						instructions += createItems(*it);
+
+				instructions.emplace_back(_pattern.instruction());
+
+				return instructions;
+			}
+			else
+				assertThrow(false, OptimizerException, "cannot create items from non-bound operation pattern without operation values");
+		case NewOptimizerPattern::Kind::Unknown:
+			assertThrow(false, OptimizerException, "cannot create items from non-bound unknown pattern");
+		default:
+			assertThrow(false, OptimizerException, "unreachable");
+	}
+}
+
+template<typename It1, typename It2>
+bool iteratorEqual(It1 _begin1, It1 _end1, It2 _begin2, It2 _end2)
+{
+	return (_end1 - _begin1 == _end2 - _begin2) && std::equal(_begin1, _end1, _begin2);
+}
+
+template<typename It, typename>
+bool dev::solidity::NewOptimizerPattern::matches(It _begin, It _end, bool _unbind)
```

Comment:
Why does it need to match a range? Why can't it just start trying to match at a certain point and then work its way backwards?